### DISCUSSION TITLE
Surprising behavior

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#2201] [Ideas / Feature Requests]
I came across this surprising behavior which was a PITA to debug.  This should be better documented somewhere if it isn't already (perhaps I missed it, wasn't in the obvious places like ui.dialog page).  Also consider improving the ui.* constructor parameters.

# Aside
Before getting to the issue itself....
Look - I'm not complaining to cause trouble.  Nicegui is great in many ways and I fervently want it to succeed.  These kind of issues may be simple or obvious if you're a nicegui expert who programs in it every day for many months.  But they are not obvious at all for beginner to intermediate users.  

Hidden gotchas are the death of software projects: death by a thousand cuts.  A few annoyances are fine; software is complex.  But each one is a drain on your users.  It drains mental energy figuring out why things don't work as described / expected.  As more and more annoyances add up, users get frustrated and move on to a different toolkit.  Incorrect or misleading documentation is worse than no documentation.

Nicegui isn't in that category yet, at least for me.  But I see signs that could get worse if left unchecked.  I give lengthy descriptions to illustrate not just what the problem is, but why it matters to users.  Hope this is received in the constructive spirit it's intended.  Text is difficult for conveying nuance.

For a positive example, compare with [BeautifulSoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/).  Docs are absolutely stellar.  I've used it happily for over a decade with nary a complaint.  It's a mature project, nicegui will get there.  Let's just keep improving.

# Sample code

Let's do something simple: a file rename dialog.  Open dialog, populate with current filename, user edits with ui.input, and we rename.  Easy peasy.

```py
async \
def rename_dialog (filename) :

    # object to hold submitted text value 
    data = { 'newfile' : '' }

    with ui.dialog () as dialog , ui.card ().classes ('dialog rename') :

        dialog.effingdone = false    # await dialog, dialog submit, dialog.set_value don't effing work

        ui.label ('Original filename')
        ui.label (filename)  # static text for reference 

        textin = ui.input (value = filename)
        textin.bind_value (data, 'newfile')

    # more code to wait for dialog, set effingdone when finished, and do rename
```

# Problem
Anyone spot the problem?  If so you've spent a lot of time learning nicegui's undocumented quirks.

Run this code (or something like it, this code is illustrative and not intended to work in your environment) and you'll spot the problem right away.  The input box is empty.  ui.input has no starting text, even though you set it with `value = filename`.

You can spend a lot of time chasing ghosts, checking whether filename is empty (nope, it isn't) or why ui.input ignores the value arg.  Finally you realize the problem isn't value = .  The problem is bind_value.  Calling bind_value overrides the value = parameter, obliterating your default text.

Why is this bad?  Several reasons:
- binding values is the "correct" or "preferred" way to handle inputs, per the nicegui docs and discussion board. Yet...
- using ui.input the "intended" way makes the value = parameter useless.  Why set a default value that will be wiped out on the very next line?
- ui.input, the most natural candidate in the world for binding a value, _doesn't_ provide a way to set a bind target in the constructor.  You have to call bind_value separately.  This isn't even mentioned or shown on the [ui.dialog page](https://nicegui.io/documentation/dialog).  A common problem for many nicegui components: critical functions to actually use the component aren't covered well, if at all, on their own doc page.
- the biggest problem is that *none of this is documented*.  You read the ui.dialog page and pass value = like it says.  Suddenly this bites you and you're scrambling to figure out why the heck it doesn't work.
- It's worse than being undocumented behavior.  This is a case where the docs actively mislead you.  Docs say one thing (pass value!  that will totally set your default!) and implementation does something radically different (yes value = "works" if you don't bind a value.  which is completely counterintuitive since then it's a PITA to get the text value back out.  the intended way should not have hidden gotchas).

# Fixes
Suggested fixes:
- Document how to actually use components like ui.input.  Few of them have binding examples on their pages.
- Don't have do-nothing parameters that won't work in practice.
- Accept bind targets in constructors.  That's how you want users to get the value out.  Why hide the ball?  Why make life difficult?
-  Document all this.  Put a big fat honking disclaimer on value = parameter that says "warning!  this parameter will dissolve into dust like a thanos snap if you bind a value like we told you to!"

Hope this helps.

**Comment by natankeddem** [comment]
I can't really speak to your points here, I am not sure how to make your "gotchas" with value binding better. What I will say in my personal opinion, binding is just another tool. To be honest value binding is something I use very rarely in my own NiceGUI projects. Maybe I don't have the imagination for implementing it efficiently? It just doesn't seem like the "right" tool for the job in many cases. If I want to access the value of an element, I'll just grab `element.value`. I find other sorts of binding like visibility more useful than value binding.

**Reply by ed2050** [reply]
It's a good point and I may move in that direction.  The downside is it requires maintaining references to all your input elements.  One or two, no problem.  When you have a dozen or more inputs for different aspects on each page, managing all those refs becomes a pain.  Anyway I'm trying the recommended way for now.

Thanks natan.

**Reply by natankeddem** [reply]
You either need to keep a reference to bind the value to or just keep a reference to the element itself. I find keeping track of the element is more useful, you can get values and do so much more. Once https://github.com/zauberzeug/nicegui/pull/1931 goes live it will make interacting with elements directly much easier without needing to maintain your references.

**Reply by ed2050** [reply]
Indeed #1931 will be great once it's ready.

I totally agree: tracking elements has many advantages.  In some situations it's the best approach.  In the example above, it's not clearly better.  ui.input only appears with the dialog, no need to do anything with it except get the value.  Either approach should work fine.

**Comment by falkoschindler** [comment]
Thanks, @ed2050, we really appreciate your feedback. But I have to object in several points:

- I don't get what all this has to do with dialogs. Why do you expect the documentation on `ui.dialog` to contain information about binding or input elements? Sure, it _could_ include an example with a `ui.input`. But I don't think one can expect that.
- We keep binding out of the initializer, because it would blow up its API drastically. Besides that we'd need to copy/paste lots of code to all elements, which is currently neatly wrapped into re-usable mixins. Like many others (`classes()`, `props()`, `style()`, `tailwind()`, ...), our binding methods follow the builder pattern, which is not uncommon and helps to structure the API.
- Your complaint seems to boil down to this:
    ```py
    data = {'newfile': ''}
    text = ui.input(value='foo')
    text.bind_value(data, 'newfile')
    ```
    When you create an input element with `value='foo'`, the element _has_ the value "foo". When you later bind it to some data, you'd expect its value to be bound to that data. I don't see a problem with that. Of course a `value` can change if it is overwritten. I don't think that we need to warn the user about it.

Although I don't agree with your concerns about value binding, I'm totally on your side regarding the bigger picture. We're eager to improve NiceGUI's usability and documentation as much as possible. We just need to find the right content for the right places.

**Reply by ed2050** [reply]
> I don't get what all this has to do with dialogs. 

You're right.  I was thinking ui.input when I wrote ui.dialog.  My bad.  Brain freeze.

> We keep binding out of the initializer, because it would blow up its API drastically. Besides that we'd need to copy/paste lots of code to all elements, which is currently neatly wrapped into re-usable mixins. Like many others (classes(), props(), style(), tailwind(), ...), our binding methods follow the builder pattern, which is not uncommon and helps to structure the API.

Maybe you're right and this is the best approach.  I don't know all your internals.  What I can say is it's confusing for users as currently presented.

> When you create an input element with value='foo', the element has the value "foo". When you later bind it to some data, you'd expect its value to be bound to that data. I don't see a problem with that. Of course a value can change if it is overwritten. I don't think that we need to warn the user about it.

The problem is you silently erase unconnected data.  Set bind to store new value input by user, silently lose initial value set in constructor.  That's unexpected behavior.

Of course you don't see a problem with it.  You designed nicegui to work that way.  The behavior isn't surprising to someone with lots of internal knowledge.

End users don't see things that way.  In one place you say "give me a string, I'll make it the default".  In another place you say "give me a location, I'll store the result there".  What you don't say is "hey that default string you gave me before - I'll throw it away and pretend it never existed".  The result is not logically connected to the premises.  It takes internal knowledge to see the connection.

# Fixes

What is documentation, if not to explain hidden connections?  What you think is incredibly obvious won't be to others without your knowledge.  Look at the python stdlib docs.  They explain all the ins and outs of calling that function, how parameters are used / ignored, and how it interacts with other functions.  That's the gold standard.

If you don't want to change the code base, that's totally fine.  I get that.  You're in far better position to judge what changes would be required, the tradeoffs, and whether it's a good approach.  I trust your decisions.

But changing documentation is trivial.  Especially when it's just adding a warning that hey, this thing may not work how you expect.  Why wouldn't you want to add a simple caveat to save users energy and time?  I don't understand the attitude.  Unless you meant something else.

**Reply by falkoschindler** [reply]
Quick question: Are you aware of the difference between `bind()`, `bind_to()` and `bind_from()`? Maybe that's the root cause of our misundertstanding. `bind` introduces two-way binding (as mentioned [here](https://nicegui.io/documentation/section_binding_properties#bindings)) and therefore has impact in both ways. If you only want to write do `data`, you can use `bind_value_to()`.

> Why wouldn't you want to add a simple caveat to save users energy and time?

I do want to. But as far as I can tell, we haven't identified what exactly needs to be changed to avoid a confusion like this. Can you point to a place in the documentation and suggest some text that would help? A pull request would also be great.

**Reply by ed2050** [reply]
> Are you aware of the difference between bind(), bind_to() and bind_from()? 

Not until just now.  I take a transactional nature to the docs.  Read through them all when I first started, but too much to remember it all.  Now I hunt for specific needs as they arise, looking at docs and forum for examples.

[This text on binding](https://nicegui.io/documentation/section_binding_properties) doesn't make it clear to me that previous values are lost / overwritten with two way binding.  It says:

> Each element provides methods like bind_value and bind_visibility to create a two-way binding with the corresponding property. To define a one-way binding use the _from and _to variants of these methods.

Two way means going forward, both directions are updated.  Doesn't say what happens to existing values.  Could just as easily discard the existing value of bound property in favor of element value, which makes more sense to me.  Either way, it should be clarified.

# Proposals

> we haven't identified what exactly needs to be changed to avoid a confusion like this. Can you point to a place in the documentation and suggest some text that would help?

Here's what I propose.  Not expecting this to be final form, feel free to improve the wording.

1. [ui.input page](https://nicegui.io/documentation/section_controls#text_input) - change value parameter description to:
```
value: the current value of the text input.  warning: if you bind or bind_from a value to this element, the current value in the element is discarded. 
```

2. same warning on any other element constructor with a parameter for inital value

3. On [binding page](https://nicegui.io/documentation/section_binding_properties), new note after first para:
```
Each element provides methods like bind_value and bind_visibility to create a two-way binding with the corresponding property. To define a one-way binding use the _from and _to variants of these methods. Just pass a property of the model as parameter to these methods to create the binding.

Note: bind and bind_from will discard the value currently in the element in favor of the new bound value
```

4. [ui.input page](https://nicegui.io/documentation/input) - on bind_value method, change description as follows (second sentence is new):
```
Bind the value of this element to the target object's target_name property.  This will discard the value currently stored in the element.
```
5. Same change to bind_value_from method of ui.input.
6. Same changes on bind_value and bind_from_value for every element that uses bind.  Presumably all are generated from the same underlying template.

**Reply by falkoschindler** [reply]
I just created PR #2323 to extend the docstrings of binding methods. Unfortunately we haven't found a simple way to generate them from a template, so I had to touch every single method.

But I don't like to extend the documentation for every bindable property like `value`. This would require changes to every element's `__init__` method for every parameter like `value`, `text`, `name`, `content`, ... I think describing the binding behavior where it actually happens (in the binding methods) is sufficient.

**Comment by J3ronimo** [comment]
@ed2050 
The whole idea that the binding should be updated from an initial constructor `value` breaks once you have multiple controls that bind to the same value:

```python
data = {'newfile': ''}
text1 = ui.input(value='foo')
text2 = ui.input(value='bar')
text1.bind_value(data, 'newfile')
text2.bind_value(data, 'newfile')
```

Since bound to the same value, text1 and text2 can never expect to stay at their different initial values. Even if it were a one way bind, they would conflict with each other. 

So the only coherent solution here can be that the bindable value itself is the single source of truth,
and that it takes control over all bound values. And that means if you want to change the initial value, you have to change it at the source:

```python
data = {'newfile': 'foo'}    #  init value
text = ui.input().bind_value(data, 'newfile')    # dont set value here, let it bind
```

I'm new to NiceGUI myself, but to me this seemed logical and clear from the get-go, so I disagree that this has to be clarified in the docs for `value`.  Maybe a short paragraph to the "Binding Properties" section could be added, including a hint where to put initial values when binding, but more than that would IMHO not be useful for the majority of people.
